codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoJmespathExpressionGenerator.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import static software.amazon.smithy.go.codegen.util.ShapeUtil.STRING_SHAPE;
+import static software.amazon.smithy.go.codegen.util.ShapeUtil.expectMember;
+import static software.amazon.smithy.go.codegen.util.ShapeUtil.listOf;
+import static software.amazon.smithy.utils.StringUtils.capitalize;
+
+import java.util.List;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.jmespath.JmespathExpression;
+import software.amazon.smithy.jmespath.ast.FieldExpression;
+import software.amazon.smithy.jmespath.ast.FunctionExpression;
+import software.amazon.smithy.jmespath.ast.ProjectionExpression;
+import software.amazon.smithy.jmespath.ast.Subexpression;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Traverses a JMESPath expression, producing a series of statements that evaluate the entire expression. The generator
+ * is shape-aware and the return indicates the underlying shape being referenced in the final result.
+ * <br/>
+ * Note that the use of writer.write() here is deliberate, it's easier to structure the code in that way instead of
+ * trying to recursively compose/organize Writable templates.
+ */
+@SmithyInternalApi
+public class GoJmespathExpressionGenerator {
+    private final GoCodegenContext ctx;
+    private final GoWriter writer;
+    private final Shape input;
+    private final JmespathExpression root;
+
+    private int idIndex = 1;
+
+    public GoJmespathExpressionGenerator(GoCodegenContext ctx, GoWriter writer, Shape input, JmespathExpression expr) {
+        this.ctx = ctx;
+        this.writer = writer;
+        this.input = input;
+        this.root = expr;
+    }
+
+    public Result generate(String ident) {
+        writer.write("v1 := $L", ident);
+        return visit(root, input);
+    }
+
+    private Result visit(JmespathExpression expr, Shape current) {
+        if (expr instanceof FunctionExpression) {
```

Comment:
[nit] Since Java 17 you can do 
```
if (expr instanceof FunctionExpression fexpr) {
    // look ma, typecasting is already done
    return visitFunction(fexpr, current)
}
```

[also nit, but tangential] if we could switch to Java 21 we could even do Pattern Matching and do something like
```
return switch (expr) {
    case FunctionExpression fexpr -> return visitFunction(fexpr, current);
    ....
}
```